### PR TITLE
BME280 SPI

### PR DESCRIPTION
**Pull request by apbodrov**
# What does this implement/fix?

Implements BME280 over SPI.

**Breaking change notes:** The original `bme280` component has been renamed to `bme280_i2c` for consistency with other platforms. To update your configuration, simply replace `bme280` with `bme280_i2c`. No other changes are required.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** implements https://github.com/esphome/feature-requests/issues/1321

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3273

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
spi:
  clk_pin: GPIO18
  mosi_pin: GPIO23
  miso_pin: GPIO19

sensor:
  - platform: bme280_spi
    temperature:
      name: "BME280 Temperature"
      oversampling: 16x
    pressure:
      name: "BME280 Pressure"
    humidity:
      name: "BME280 Humidity"
    update_interval: 60s
    cs_pin: GPIO5
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by apbodrov** [comment]
tested two BMEs with config

```yaml
spi:
  clk_pin: GPIO18
  mosi_pin: GPIO23
  miso_pin: GPIO19

i2c:
  sda: 21
  scl: 22
  scan: true

sensor:
  - platform: bme280_spi
    temperature:
      name: "BME280 SPI Temperature"
      oversampling: 16x
    pressure:
      name: "BME280 SPI Pressure"
    humidity:
      name: "BME280 SPI Humidity"
    update_interval: 60s
    cs_pin: GPIO5


  - platform: bme280
    temperature:
      name: "BME280 I2C Temperature"
      oversampling: 16x
    pressure:
      name: "BME280 I2C Pressure"
    humidity:
      name: "BME280 I2C Humidity"
    update_interval: 60s
    address: 0x76
```

**Comment by apbodrov** [comment]
@OttoWinter @jesserockz hello! can you check my PR? Thanks!

**Comment by jesserockz** [comment]
Please don't ping maintainers. Its a good way to get ignored :wink:

**Comment by apbodrov** [comment]
@jesserockz 
I'm following your **_own_** guide ;)
изображение

**Comment by jesserockz** [comment]
3 days is not lingering yet. It's still fresh out of the oven.

**Comment by rossens** [comment]
I am looking forward to trying/using this - I have several SPI BME280 and Pi Pico ready and waiting. I'm fairly new to ESPHome - can anyone advise how long it usually takes for something like this to make it over to Home Assistant?
PS - seems one of the Validate YAML tests failed - "Pin 5 is used in multiple places."?

**Comment by kbx81** [comment]
@rossens you (or anyone else who might be interested) can test this PR as an external component by adding this to your yaml config:
```yaml
external_components:
  - source: github://pr#5538
    components: [ bme280_base, bme280_spi ]
```

**Comment by kbx81** [comment]
While it means a breaking change, for consistency with other components within the platform where the hardware (sensor, display, etc.) is able to support both I2C and SPI, the I2C component should probably be renamed to `bme280_i2c`.

**Comment by apbodrov** [comment]
> While it means a breaking change, for consistency with other components within the platform where the hardware (sensor, display, etc.) is able to support both I2C and SPI, the I2C component should probably be renamed to `bme280_i2c`.

In fact, the only difference will be another name, but it will take a lot of work to

1) Refactor the code, tests, and documentation in the repository.
2) Refactor the _alll_ configs of all users.

I also think that bme280_i2c sounds more logical, but there is not much point in renaming it.

**Comment by rossens** [comment]
> @rossens you (or anyone else who might be interested) can test this PR as an external component by adding this to your yaml config:
> 
> ```yaml
> external_components:
>   - source: github://pr#5538
>     components: [ bme280_base, bme280_spi ]
> ```

Thanks that looks much simpler than what I did (I applied this PR as a patch over my installed ESPHome files). I got it working fine though. Will try this better way next time.

Results: Temperature and pressure seem pretty close - but humidity is reading too low. e.g. currently 25% reported vs 50% actual. 
Wondering if it could be anything to do with the PR - but assume it is more likely an issue with the cheap sensors I purchased from AliExpress.
 
I also wonder if it could be relative vs absolute humidity? I am assuming the sensor/code and ESPHome all talk "relative" but I'm not 100% sure. Anyone know?

I will continue to investigate on my end and advise if/when I get correct Humidity - what was the issue.

**Comment by apbodrov** [comment]
@rossens 
> Results: Temperature and pressure seem pretty close - but humidity is reading too low. e.g. currently 25% reported vs 50% actual. Wondering if it could be anything to do with the PR - but assume it is more likely an issue with the cheap sensors I purchased from AliExpress.

Hello! I had a similar problem, and I also decided that it was the sensor. But since I'm not the only one with the problem, now I'm starting to think that the problem is in the code.

I'll try to fix the bug and synchronize the readings later.

Thanks!

**Comment by rossens** [comment]
I'm suspicious the problem may be here:
```
float BME280Component::read_humidity_(const uint8_t *data, int32_t t_fine) {
  uint16_t const raw_adc = ((data[6] & 0xFF) << 8) | (data[7] & 0xFF); // <<<<<<<<<<<<<<<
```
Depending on compiler - `(data[6] << 8)` may always give 0 because `uint8_t(x) << 8 = uint8_t(0)`

Try this instead... (I also removed the `& 0xFF` which aren't neccesary)
```
  uint16_t const raw_adc = (uint16_t(data[6]) << 8) | data[7];
```

I'll also give this a try later on today myself.

**Comment by apbodrov** [comment]
> I'm suspicious the problem may be here:
> `float BME280Component::read_humidity_(const uint8_t *data, int32_t t_fine)`

This is a shared code between I2C and SPI component, it was moved from BME280 I2C. So that fix will affect I2C code (which I do not touched, just move)

**Comment by rossens** [comment]
> I'm suspicious the problem may be here:
> `float BME280Component::read_humidity_(const uint8_t *data, int32_t t_fine)`

I just tried this out - no improvement in humidity reading - so my suspicion wasn't correct.

File: esphome/components/bme280_spi/bme280_spi.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  this->delegate_->read_array(data, len);
+  this->disable();
+  return true;
+}
+
+bool BME280SPIComponent::read_byte_16(uint8_t a_register, uint16_t *data) {
+  this->enable();
+  this->delegate_->transfer(set_bit(a_register, 7));
+  data[1] = this->delegate_->transfer(0);
+  data[0] = this->delegate_->transfer(0);
```
Comment (rossens) [inline esphome/components/bme280_spi/bme280_spi.cpp]: Looks like the humidity issue is coming from here...
`data` is `uint16_t*` - but code looks like `u8*` was intended... e.g.
```
  ((uint8_t*)data)[1] = this->delegate_->transfer(0);
  ((uint8_t*)data)[0] = this->delegate_->transfer(0);
```
I tried the above fix and my humidity increased from 35% to 50% - which I "hope" is now correct.

Comment (apbodrov) [reply esphome/components/bme280_spi/bme280_spi.cpp]: Thank you! Will fix with tests shortly

**Review by jesserockz — DISMISSED** [review]
As per the message above from @kbx81

> While it means a breaking change, for consistency with other components within the platform where the hardware (sensor, display, etc.) is able to support both I2C and SPI, the I2C component should probably be renamed to `bme280_i2c`.

Please move the main component. This is to avoid confusion in the future and we have done it many times in the past.

File: CODEOWNERS [hunk 2]
```diff
@@ -54,6 +54,8 @@ esphome/components/bl0940/* @tobias-
 esphome/components/bl0942/* @dbuezas
 esphome/components/ble_client/* @buxtronix
 esphome/components/bluetooth_proxy/* @jesserockz
+esphome/components/bme280_base/* @esphome/core
+esphome/components/bme280_spi/* @apbodrov
```
Comment (rossens) [inline CODEOWNERS:58]: Is it an issue this says `Unknown owner: make sure @apbodrov exists and has write access to the repository`

File: esphome/components/bme280_spi/bme280_spi.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+// In SPI mode, only 7 bits of the register addresses are used; the MSB of register address is not used
+// and replaced by a read/write bit (RW = ‘0’ for write and RW = ‘1’ for read).
+// Example: address 0xF7 is accessed by using SPI register address 0x77. For write access, the byte
+// 0x77 is transferred, for read access, the byte 0xF7 is transferred.
+// https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bme280-ds002.pdf
+
+bool BME280SPIComponent::read_byte(uint8_t a_register, uint8_t *data) {
+  this->enable();
+  // cause: *data = this->delegate_->transfer(tmp) doesnt work
```
Comment (rossens) [inline esphome/components/bme280_spi/bme280_spi.cpp:33]: Probably my fault - but I don't understand this comment.

**Review by rossens — APPROVED** [review]
Looks good to me. 
I'm successfully using this PR/patch to operate 4x BME280 SPI sensors on a Pi Pico.

File: CODEOWNERS (hunk 2 above)
Comment (apbodrov) [reply CODEOWNERS:58]: > Is it an issue this says `Unknown owner: make sure @apbodrov exists and has write access to the repository`

Interesting, what nickname should be here in that case

File: esphome/components/bme280_spi/bme280_spi.cpp (hunk 3 above)
Comment (apbodrov) [reply esphome/components/bme280_spi/bme280_spi.cpp:33]: > Probably my fault - but I don't understand this comment.

This method should return the read value, but it did not work properly, I had to split the line
`*data = this->delegate_->transfer(set_bit(a_register, 7))`
into two
`this->delegate_->transfer(set_bit(a_register, 7));`
`*data = this->delegate_->transfer(0);`

File: CODEOWNERS (hunk 2 above)
Comment (jesserockz) [reply CODEOWNERS:58]: This is fine and not actually a problem and can be ignored

**Comment by TCB13** [comment]
Any updates on merging this? Thank you.

**Review by kbx81 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**